datalad_container/containers_add.py

```diff
@@ -187,6 +183,12 @@ def __call__(name, url=None, dataset=None, call_fmt=None, image=None):
                     docker_image, image)
                 check_call(["docker", "pull", docker_image])
                 docker.save(docker_image, image)
+            elif op.exists(url):
+                lgr.info("Copying local file %s to %s", url, image)
+                image_dir = op.dirname(image)
+                if image_dir and not op.exists(image_dir):
+                    os.makedirs(image_dir)
+                copyfile(url, image)
```

Comment:
- possibly technically irrelevant since annex doesn't store even executable permission afaik for annexed files, but might manifest somehow many in in "unlocked v6 mode" -- `copyfile` doesn't copy permissions (image could be executable as e.g. in the case with singularity). `.copy2` seems to do `copyfile` and then `copystat`
- (could be a separate PR if too involved) is there a way to facilitate efficient copy on CoW filesystems (BTRFS)?    `cp --reflink=auto` on GNU systems, `cp -c` on OSX  achieves that in cmdline. For python quick google lead to a alpha/stale effort https://github.com/giampaolo/zerocopy which I didn't check inside (may be just does those cmdline calls?)

Reply:
> * possibly technically irrelevant since annex doesn't store even executable permission afaik for annexed files, but might manifest somehow many in in "unlocked v6 mode" -- `copyfile` doesn't copy permissions (image could be executable as e.g. in the case with singularity). `.copy2` seems to do `copyfile` and then `copystat`

Hmm, my thinking was that *not* preserving the executable bit was (1) consistent with the addurl behavior on the local file and (2) makes people configure `cmdexec` rather than rely on the executable bit, which may not be preserved across machines, at least until

https://github.com/datalad/datalad-container/blob/1983d6fbb3f76cf73e31b17eca5a27207e00eb2e/datalad_container/containers_add.py#L198-L199

I'm not strongly tied to the reasoning though (particularly since I view the `--url=<local-file>` behavior as mainly useful for local testing anyway), so I'm OK updating it to `copy2` if you'd prefer.

> * (could be a separate PR if too involved) is there a way to facilitate efficient copy on CoW filesystems (BTRFS)? [...]

Yeah, that seems outside the scope of this PR.

Reply:
Makes sense